### PR TITLE
42auth skeleton implemented

### PR DESCRIPTION
**Pull request by taoramo**
You can test 42auth at https://localhost/login.html and clicking the button which should redirect you to the 42 login page. Eventually this button will be on index.html. 

What should happen after succesful login is that a new user with fetched display name is added to database and the popup closed. This works only if the user doesn't exist already in the database. As a stupid hack for testing I've just been adding letters to the stored user and displayname at views.py:110 and views.py:113. The data retrieved from 42 includes uri to profile pic etc, to be implemented. There's probably also some error handling to do.

Also I added a modules.txt to root of repo to keep track of what everyone's doing. Feel free to make it better if you like.

### CODE REVIEW

File: backend/players/views.py [hunk 1]
```diff
@@ -61,4 +72,85 @@ def create_player(request):
...
+             return JsonResponse({'error': 'No code provided'}, status=400)
+        token_response = exchange_code_for_token(code)
+        if 'error' in token_response:
+            return JsonResponse(token_response, status=400)
+        user_data = fetch_42_user_data(token_response['access_token'])
+        if 'error' in user_data:
+            return JsonResponse(user_data, status=400)
+        try:
+            # Create the user
+            user = User.objects.create_user(username=f'{user_data['login']}')
```
Comment (madfcat) [inline backend/players/views.py]: As I understand, single quote in single quotes should not be a problem starting from Python 3.12.
Still the line can be simplified to: 
```suggestion
            user = User.objects.create_user(username=user_data['login'])
```

File: backend/players/views.py [hunk 2]
```diff
@@ -61,4 +72,85 @@ def create_player(request):
...
+            # Create the player with the associated user
+            player = Player.objects.create(
+                user=user, display_name=f'{user_data['displayname']}')
```
Comment (madfcat) [inline backend/players/views.py]:
```suggestion
                user=user, display_name=user_data['displayname'])
```

File: backend/players/views.py [hunk 3]
```diff
@@ -61,4 +72,85 @@ def create_player(request):
...
+        'client_secret': os.environ.get('OAUTH_CLIENT_SECRET'),
+        'code': code,
+        'redirect_uri': os.environ.get('OAUTH_REDIRECT'),
+       }
+ try:
+        response = requests.post(token_url, data=data)
+        response.raise_for_status()
+        return response.json()
+ except requests.exceptions.RequestException as e:
+        return {'error': 'Token exchange failed: {str(e)}'}
```
Comment (madfcat) [inline backend/players/views.py]:
```suggestion
        return {'error': f'Token exchange failed: {str(e)}'}
```

**Review by madfcat — CHANGES_REQUESTED** [review]
I tested the solution (created app in intra and added these envs):
```
OAUTH_CLIENT_ID=
OAUTH_CLIENT_SECRET=
OAUTH_REDIRECT=http://localhost:8000/players/oauth/callback/
```

It worked for me with creating user and redirection

In our endpoints, we will use `@login_required` decorator for checking if user is logged in to authorize the usage of the endpoint. This decorator checks if the session cookie was provided.

As I understand, `Set-Cookie` with session and csrf_token as well (we should implement it later) will be added to the response if we use `login` function in our authentication logic on the Django backend.
We will not use templates, so the template logic should be removed.

For example this part:
```
        try:
            # Create the user
            user = User.objects.create_user(username=f'{user_data['login']}')
            # Create the player with the associated user
            player = Player.objects.create(
                user=user, display_name=f'{user_data['displayname']}')

            # Return success response
            # return HttpResponseRedirect('http:://localhost/callback.html')
            template = get_template('callback.html')  # or 'your_app/callback.html'
            return HttpResponse(template.render({}, request))

        except Exception as e:
            return JsonResponse({'error': str(e)}, status=500)
        except TemplateDoesNotExist:
        # If template is not found, return error with helpful debugging info
            return HttpResponse(f"""
            Template not found. Debug info:
            - App directory: {__file__}
            - Available templates: {[t.name for t in get_template().engine.get_template_sources('callback.html')]}
        """)
```

Could be changed to something like:
```
        try:
            # Check if user exists or create a new user
            user, created = User.objects.get_or_create(username=user_data['login'])
            if created:
                # Create the Player model instance for this user
                player = Player.objects.create(user=user, display_name=user_data['displayname'])

            # Authenticate and log the user in
            user = authenticate(request, username=user.username)
            if user is not None:
                login(request, user)  # Log the user in and create session
                # Redirect or return a success response to the frontend
                return JsonResponse({'message': 'Successfully authenticated', 'id', user.player.id, 'username': user.username, 'displayName': user.player.display_name})

            else:
                return JsonResponse({'error': 'Authentication failed'}, status=400)

        except Exception as e:
            return JsonResponse({'error': str(e)}, status=500)
```

Also, please consider refactoring added in suggestions and one little typo where the exception is caught.

I think we should use Django built login to return the session and csrf_token to make it secure.
Then we can store the token in the frontend to avoid CSRF attacks.
`csrf_token` and `sessionId` are included in the cookies sent from server when we use `login()` function in the endpoint. For example, like this:
`login(request, user)`

Please, can you do this change?

**Comment by taoramo** [comment]
Changing file structure a bit getting rid of templates so opnening a new pull request.

**Pull request closed without merging**